nemo/collections/asr/modules/rnnt.py

```diff
@@ -50,6 +50,525 @@
+        embeds = []
+        for i in range(self.context_size):
+            # We use different embedding matrices for different context positions.
+            # In this list, a smaller index means more recent history word.
+            # We assign more dimensions for the most recent word in the history.
+            if i != 0:
+                embed_size = emb_dim // 2 // self.context_size
+            else:
+                embed_size = emb_dim - (emb_dim // 2 // self.context_size) * (self.context_size - 1)
+
+            if blank_as_pad:
+                embed = torch.nn.Embedding(vocab_size + 1, embed_size, padding_idx=blank_idx)
+            else:
+                embed = torch.nn.Embedding(vocab_size, embed_size)
+
+            embeds.append(embed)
+
+        self.embeds = torch.nn.ModuleList(embeds)
+        self.blank_idx = blank_idx
+
+    def forward(self,
+                y: Optional[torch.Tensor] = None,
+                state: Optional[List[torch.Tensor]] = None,
+                ):
+        # Although this is a *stateless* net, we use the "state" parameter to
+        # pass in the previous labels, unlike LSTMs where state would represent
+        # hidden activations of the network.
+        # "state" is a list of 1 tensor in order to be consistent with the stateful
+        # decoder interface, and the element is a tensor of shape [batch, context-length].
+        # The out dimension of this function is B x U x D, with D being the total embedding dim.
+        outs = []
+
+        [B, U] = y.shape
+        appended_y = y
+        if state != None:
+            appended_y = torch.concat([state[0], y], axis=1)
+            context_size = appended_y.shape[1]
+
+            if context_size < self.context_size:
+                padded_state = torch.ones([B, self.context_size], dtype=torch.long, device=y.device) * self.blank_idx
+                padded_state[:,self.context_size - context_size:] = appended_y
+            elif context_size == self.context_size + 1:
+                padded_state = appended_y[:,1:]
+            else:
+                padded_state = appended_y
+
+            for i in range(self.context_size):
+                out = self.embeds[i](padded_state[:,self.context_size - 1 - i:self.context_size - i])
+                outs.append(out)
+        else:
+            for i in range(self.context_size):
+                out = self.embeds[i](y)
+
+                if i != 0:
+                    out[:,i:,:] = out[:,:-i,:].clone()  # needs clone() here or it might complain about src and dst mem location have overlaps.
+                    out[:,:i,:] *= 0.0
+                outs.append(out)
+
+        out = self.dropout(torch.concat(outs, axis=-1))
+        out = self.norm(out)
+
+        state = None
+        if y is not None:
+            state = [appended_y[:,appended_y.shape[1] - self.context_size + 1:]]
+        return out, state
+
+
+class RNNTStatelessDecoder(rnnt_abstract.AbstractRNNTDecoder, Exportable):
```

Comment:
Its.. not really an RNN anymore is it? Or is it vaguely recurrent in the sense that while we don't have true state of an rnn, we do have the embedding history itself as state.. guess its fine to name it this.

Better name is just "StatelessDecoder" or "StatelessTransducerDecoder"

Reply:
Done